plugins/automation/app/controllers/discourse_automation/admin_automations_controller.rb

```diff
@@ -37,69 +37,34 @@ def show
     def create
       automation_params = params.require(:automation).permit(:script, :trigger)
 
-      automation =
-        DiscourseAutomation::Automation.new(
-          automation_params.merge(last_updated_by_id: current_user.id),
-        )
-
-      if automation.scriptable&.forced_triggerable
-        automation.trigger = automation.scriptable.forced_triggerable[:triggerable].to_s
+      DiscourseAutomation::Create.call(params: automation_params.to_h, guardian:) do
+        on_success { |automation:| render_serialized_automation(automation) }
+        on_failed_policy(:can_create_automation) { raise Discourse::InvalidAccess }
+        on_failed_contract do |contract|
+          render json: failed_json.merge(errors: contract.errors.full_messages),
+                 status: :bad_request
+        end
+        on_failure { raise Discourse::InvalidParameters }
       end
-
-      automation.save!
-
-      render_serialized_automation(automation)
     end
 
     def update
-      params.require(:automation)
-
-      automation =
-        DiscourseAutomation::Automation.includes(:fields, :pending_automations).find(params[:id])
-
-      automation.perform_required_fields_validation = true
-
-      if automation.scriptable.forced_triggerable
-        params[:trigger] = automation.scriptable.forced_triggerable[:triggerable].to_s
-      end
-
-      attributes =
-        request.parameters[:automation].slice(:name, :id, :script, :trigger, :enabled).merge(
-          last_updated_by_id: current_user.id,
-        )
-
-      if attributes.key?(:trigger) && automation.trigger != params[:automation][:trigger]
-        params[:automation][:fields] = []
-        attributes[:enabled] = false
-        automation.fields.destroy_all
-      end
-
-      if attributes.key?(:script)
-        if automation.script != params[:automation][:script]
-          attributes[:trigger] = nil
-          params[:automation][:fields] = []
-          attributes[:enabled] = false
-          automation.fields.destroy_all
-          automation.tap { |r| r.assign_attributes(attributes) }.save!(validate: false)
-        else
-          Array(params[:automation][:fields])
-            .reject(&:empty?)
-            .each do |field|
-              automation.upsert_field!(
-                field[:name],
-                field[:component],
-                field[:metadata],
-                target: field[:target],
-              )
-            end
-
-          automation.update!(attributes)
+      automation_params = params.require(:automation).permit(:name, :script, :trigger, :enabled)
+      fields = params[:automation][:fields]&.map(&:to_unsafe_h)
+
+      DiscourseAutomation::Update.call(
+        params: automation_params.to_h.merge(automation_id: params[:id], fields: fields),
```

Comment:
Since the contract will filter out unwanted parameters, we can actually do this:
```suggestion
        params.require(:automation).to_unsafe_h.merge(automation_id: params[:id])
```